src/cargo/core/compiler/job_queue.rs

```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+        }
+
+        Ok(())
+    }
+
+    fn grant_rustc_token_requests(&mut self) -> CargoResult<()> {
+        // If we managed to acquire some extra tokens, send them off to a waiting rustc.
+        let extra_tokens = self.tokens.len() - (self.active.len() - 1);
+        for _ in 0..extra_tokens {
+            if !self.to_send_clients.is_empty() {
+                // remove from the front so we grant the token to the oldest
+                // waiter
+                let (id, client) = self.to_send_clients.remove(0);
+                let token = self.tokens.pop().expect("an extra token");
+                self.rustc_tokens
+                    .entry(id)
+                    .or_insert_with(Vec::new)
+                    .push(token);
+                client
+                    .release_raw()
+                    .chain_err(|| "failed to release jobserver token")?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn handle_event(
+        &mut self,
+        cx: &mut Context<'a, '_>,
+        jobserver_helper: &HelperThread,
+        plan: &mut BuildPlan,
+        event: Message,
+    ) -> CargoResult<Option<anyhow::Error>> {
+        match event {
+            Message::Run(id, cmd) => {
+                cx.bcx
+                    .config
+                    .shell()
+                    .verbose(|c| c.status("Running", &cmd))?;
+                self.timings.unit_start(id, self.active[&id]);
+            }
+            Message::BuildPlanMsg(module_name, cmd, filenames) => {
+                plan.update(&module_name, &cmd, &filenames)?;
+            }
+            Message::Stdout(out) => {
+                cx.bcx.config.shell().stdout_println(out);
+            }
+            Message::Stderr(err) => {
+                let mut shell = cx.bcx.config.shell();
+                shell.print_ansi(err.as_bytes())?;
+                shell.err().write_all(b"\n")?;
+            }
+            Message::FixDiagnostic(msg) => {
+                self.print.print(&msg)?;
+            }
+            Message::Finish(id, artifact, result) => {
+                let unit = match artifact {
+                    // If `id` has completely finished we remove it
+                    // from the `active` map ...
+                    Artifact::All => {
+                        info!("end: {:?}", id);
+                        self.finished += 1;
+                        if let Some(rustc_tokens) = self.rustc_tokens.remove(&id) {
+                            // This puts back the tokens that this rustc
+                            // acquired into our primary token list.
+                            //
+                            // FIXME: this represents a rustc bug: it did not
+                            // release all of its thread tokens but finished
+                            // completely.
+                            self.tokens.extend(rustc_tokens);
+                        }
+                        while let Some(pos) =
```

Comment:
I'm a little worried about this loop, it feels like on larger machines it could cause a lot of shifting which could be problematic. I'll leave some comments elsewhere I think may help with this.